### DISCUSSION TITLE
message line at the bottom of Ctrlr panel

### DISCUSSION THREAD

**Discussion by synthmutt** [RomanKubiak/ctrlr#556] [Q&A] [answered]
When I change a panel's MIDI device, a message is displayed at the bottom of the panel, which goes away after a while. Can this message feature be invoked for custom messages from the panel I'm creating, or is it reserved for use by Ctrlr?

For example, I'd rather not launch a popup box saying "File loaded" and force the user to click Ok to dismiss that ... showing an informational message at the bottom that goes away without user intervention sounds nicer

**Comment by damiensellier** [comment, marked as answer]
Those messages are coming from the JUCE device change callback and unfortunately you cannot have access to it from the panel. 

But you can just add a stylized uiLabel (right click, add component, static > uiLabel)  at the bottom of the panel and display the info you want with LUA. you can Show/ hide this component after a few second etc
Can you please tell us what do you want to display and when?

**Reply by synthmutt** [reply]
Ok thanks. I'm thinking about information-only messages in the case of a successful operation, such as "File loaded" or  "Program copied".  If an error occurred, such as the selected file could not be loaded, maybe that should have a popup box ... but if everything is working fine, I'm not a fan of using a popup box which the user has to dismiss to continue their workflow.

When I change a MIDI device, Ctrlr gives a green message at the bottom of the panel saying that it opened Ok ... this message goes away after a couple seconds.  I was looking to reuse that if possible; sounds like that is not available for my own custom messages

**Reply by damiensellier** [reply]
Are you going to use some kind of display to show parameter values etc? On my panels I use the main display to show those kind of info (ready, loaded, errors etc) that reflects the external hardware current state.  
You should try to avoid  alertbox for that for the UX.
![status LCD](https://github.com/RomanKubiak/ctrlr/assets/55164385/9a145fa1-8acc-459f-859b-fd2e2218656f)

The lua code to display text in a uiLabel component is the following : 

`panel:getLabel("uiLabelName"):setText("YOUR TEXT")`

**Comment by dnaldoog** [comment]
I often have a grey strip at the bottom of a panel which is a kind of pseudo status bar. I then write a value to the status bar and have it disappear after a certain time using a timer.

See attached panel, which also has code to extract the names of combo `uiComboContent`.

<pre>
myFunction=function(mod,value,source)
    panel:getLabelComponent("sbar"):setText("Loaded: " .. value)
    timer:setCallback(199, TimerCallback) -- reset status bar
    timer:startTimer(199,2000) -- Timer 2000ms
end	
	
TimerCallback = function(timerId)
    if timerId == 199 then --reset the status bar message
        panel:getLabelComponent("sbar"):setText("")
        timer:stopTimer(timerId)
    end
end
</pre>

![sbar](https://github.com/RomanKubiak/ctrlr/assets/76831683/004830d5-93f7-44e1-94e8-a3dcf122aec7)

[message line at the bottom of Ctrlr panel_2_0.zip](https://github.com/RomanKubiak/ctrlr/files/11570233/message.line.at.the.bottom.of.Ctrlr.panel_2_0.zip)

**Reply by damiensellier** [reply]
Hi John, That's exactly what I was thinking. It won't be hard to implement.
The component can be also hidden at the end of the timer with : 

```
myNewValue = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)
    panel:getLabelComponent("sbar"):setText("Loaded: " .. value)
    panel:getLabelComponent("sbar"):setProperty("componentVisibility", 1, false) -- Display bar
    timer:setCallback(199, TimerCallback) -- reset statusbar
    timer:startTimer(199,2000) -- Timer 2000ms
end
```

AND

```
selectSomething = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)
    local contents = mod:getComponent():getProperty("uiComboContent")
    local t = split(contents)
    table.concat(t, "--")
    panel:getLabelComponent("sbar"):setText("Loaded: " .. t[value + 1])
    timer:setCallback(199, TimerCallback) -- reset statusbar
    timer:startTimer(199, 10000) -- Timer 10000ms
end

TimerCallback = function(timerId)
    if timerId == 199 then --reset the status bar message
        panel:getLabelComponent("sbar"):setText("")
        panel:getLabelComponent("sbar"):setProperty("componentVisibility", 0, false) -- hide bar
        timer:stopTimer(timerId)
    end
end

split = function(str) -- split text by \n
    lines = {}
    for s in str:gmatch("[^\r\n]+") do
        table.insert(lines, s)
    end
    return lines
end
```

**Reply by dnaldoog** [reply]
That's great Damien.

I noticed in the code you pasted I had a `table.concat(t, "--")` which was part of my debugging, so I removed that and uploaded a new panel in the post above!

**Comment by dnaldoog** [comment]
Crickets

**Reply by damiensellier** [reply]
ahaha just passing by :) Have a great weekend John!

**Reply by dnaldoog** [reply]
> ahaha just passing by :) Have a great weekend John!

Yes, just passing by! Hope all is well for you! 😔
